### ISSUE TITLE
JSON Editor var expression green/white color issue

### ISSUE THREAD

**Issue by 96LawDawg** [ArnoldSmith86/virtualtabletop#1447]
When working on my new functions for color, I noticed a color issue in the JSON Editor.  For example:
`"var testMe = colorMono '#5c5005'"` shows the hex value as white when it should(?) be green.  In fact, any symbol not a number or letter right after the single quote does that. I don't know if you would ever use something like `'!hello'` or `'*nice to see you'` in a valid expression, but at least the # sign in a hex value should show up as green, I think, given where the color functions are headed.  Probably a regex somewhere that needs a small tweak.

For what it's worth, it may be line 1455 in jsonedit.js:

`[ /^( +"var )(.*)( = )(-?[0-9.]+)?(null|true|false)?(\$\{[^}]+\})?('(?:[a-zA-Z0-9,.() _-]|\\\\u[0-9a-fA-F]{4})*')?( )?([0-9a-zA-Z_-]+|[=+*/%<!>&|-]{1,3})?(🧮(?:[0-9a-zA-Z_-]+|[=+*/%<!>&|-]{1,2}))?( )?(-?[0-9.]+)?(null|true|false)?(\$\{[^}]+\})?('(?:[a-zA-Z0-9,.() _-]|\\\\u[0-9a-fA-F]{4})*')?( )?(-?[0-9.]+)?(null|true|false)?(\$\{[^}]+\})?('(?:[a-zA-Z0-9,.() _-]|\\\\u[0-9a-fA-F]{4})*')?( )?(-?[0-9.]+)?(null|true|false)?(\$\{[^}]+\})?('(?:[a-zA-Z0-9,.() _-]|\\\\u[0-9a-fA-F]{4})*')?(.*)(",?)$/, 'default', 'custom', null, 'number', 'null', 'variable', 'string', null, null, 'variable', null, 'number', 'null', 'variable', 'string', null, 'number', 'null', 'variable', 'string', null, 'number', 'null', 'variable', 'string', null, 'default' ],`

Anybody that can figure that out is a regex deity.

**Comment by robartsd** [comment]
Probably a carryover from before we whitelisted a bunch of symbols between \u0020 and \u00ff that previously required escape codes.

**Comment by rogerl50** [comment]
Group 15 in that RE is `?('(?:[a-zA-Z0-9,.() _-]|\\\\u[0-9a-fA-F]{4})*')` (and so is group 25). These seem intended to recognize character strings containing only alphanumerics plus escaped Unicode. Such strings will be green in the output. The problem case doesn't match that string, so it instead matches group 26, `?(.*)`, which is formatted in white.

It seems pretty straightforward to modify those two REs to allow other characters, but I'm not sure exactly what we would want to do there. Do we want to allow just hex values? How about other strings beginning with `#`? How about strings beginning with other characters?

Or we could just leave the whole thing alone and move on with our lives.

**Comment by robartsd** [comment]
> Group 15 in that RE is `?('(?:[a-zA-Z0-9,.() _-]|\\\\u[0-9a-fA-F]{4})*')` (and so is group 25). These seem intended to recognize character strings containing only alphanumerics plus escaped Unicode. Such strings will be green in the output. The problem case doesn't match that string, so it instead matches group 26, `?(.*)`, which is formatted in white.
> 
> It seems pretty straightforward to modify those two REs to allow other characters, but I'm not sure exactly what we would want to do there. Do we want to allow just hex values? How about other strings beginning with `#`? How about strings beginning with other characters?
> 
> Or we could just leave the whole thing alone and move on with our lives.

I think that at the time this was written, these were the only characters allowed unescaped, but we expanded the allowable characters later (PR #645).

Group 15 RE should be `('(?:[ !#-&(-[\\]-~]|\\\\u[0-9a-fA-F]{4})*')` to match any printable ASCII character other than `'`, `"`, and `\`.

**Comment by ArnoldSmith86** [comment]
I'm not 100% sure that's right but I think it's just a catch-all for anything not recognized. Or maybe to allow for `//` comments after the expression.

**Comment by rogerl50** [comment]
BTW https://extendsclass.com/regex-tester.html did a great job of interpreting that RE so it could be understood by normal humans.

**Comment by rogerl50** [comment]
And group 20 seems to be another copy of  the same RE.

**Comment by robartsd** [comment]
> BTW https://extendsclass.com/regex-tester.html did a great job of interpreting that RE so it could be understood by normal humans.

Note that the extra backslash escapes should be removed before input into this tool: `('(?:[ !#-&(-[\]-~]|\\u[0-9a-fA-F]{4})*')`

The pattern should match the same as this pattern which is slightly easier to understand: `('(?:(?=[^\\'"])[ -~]|\\u[0-9a-fA-F]{4})*')`

**Comment by rogerl50** [comment]
Groups 7, 15, 20, and 25 are all identical and equal to `?('(?:[a-zA-Z0-9,.() _-]|\\\\u[0-9a-fA-F]{4})*')`. I hope someone knows what these four groups match; should they all be changed as @robartsd recommends?

**Comment by robartsd** [comment]
Yes, I believe there should be 4 opportunities to match parameters (once before and three times after matching an operation) and each parameter could match a string literal. It would probably be easier to follow if it were structured the same way it is in widget.js:

```
        const identifier = '(?:[a-zA-Z0-9_-]|\\\\u[0-9a-fA-F]{4})+';
        const string     = `'((?:[ !#-&(-[\\]-~]|\\\\u[0-9a-fA-F]{4})*)'`;
        const number     = '(-?(?:0|[1-9][0-9]*)(?:\\.[0-9]+)?)';
        const variable   = `(\\$\\{[^}]+\\})`;
        const parameter  = `(null|true|false|\\[\\]|\\{\\}|${number}|${variable}|${string})`;

        const left       = `var (\\$)?(${identifier})(?:\\.(\\$)?(${identifier}))?`;
        const operation  = `${identifier}|[=+*/%<!>&|-]{1,3}`;

        const regex      = `^${left} += +(?:${parameter}|(?:${parameter} +)?(🧮)?(${operation})(?: +${parameter})?(?: +${parameter})?(?: +${parameter})?)(?: *|(?: +//.*))?`;

        const match = a.match(new RegExp(regex + '\x24')); // the minifier doesn't like a "$" here
``
